### ISSUE TITLE
Call C++ function using ctypes py_object argument on windows doesn't work

### ISSUE THREAD

**Issue by gtwpro** [python/cpython#114754]
# Bug report

### Bug description:

I get an access violation.  I am using a c++ dll with a single function that tries to extract a string from a PyObject.  This is from similar examples posted in stackoverflow.  Passing a c_char_p (to a function taking a c_char_p) and returning a PyObject * works fine.  Normal c_char_p and c_int work fine.

Load a C++ dll and call a function with an object.  Set correct path.

Using msvc 2022 and Python 3.9.

Python (using PYFUNCTYPE, WINFUNCTYPE or CFUNCTYPE makes no difference):
```
import ctypes
import sys
import os 

y = __file__
str_path = os.path.dirname(os.path.realpath("..\\..\\string\\x64\\Debug\\string.dll"))
handle = ctypes.CDLL(str_path + "/string.dll")     

handle.My_Function3.argtypes = [ctypes.py_object]
handle.My_Function3.restype = ctypes.py_object

z = handle.My_Function3(y)

print(z)

```
Create C++ x64 string.dll adding a single function.  I have verified the address is correctly passed through, but then I get internal errors inside Python code when calling the PyArg_* functions.  Nowhere does the documentation state that any state setup is needed to use these functions in this situation.

C++ file added to string.dll (added to Windows dll) dllmain.cop is unchanged:

function.cpp:
```C++
#include "pch.h"
#define PY_SSIZE_T_CLEAN
#include <Python.h>

char s[100];

static PyObject *MyError;

extern "C"
{
	__declspec(dllexport) PyObject* __stdcall My_Function3(PyObject *args)
	{
		const char* str;
		char s[100];

        // Code fails here with access violation
	    if (!PyArg_ParseTuple(args, "s", &str))
			return NULL;
 
		if (strlen(str) > 100)
		{
			PyErr_SetString(MyError, "Error: string too long");
			return NULL;
		}

		strcat_s(s, 100, str);
		return Py_BuildValue("s", s);
	}
}

```

### CPython versions tested on:

3.9

### Operating systems tested on:

Windows

**Comment by ronaldoussoren** [comment]
You probably want to use `PyDLL` instead of `CDLL`, see https://docs.python.org/3/library/ctypes.html#ctypes.PyDLL.

The functions exposed through `CDLL` release the GIL while calling into C code and don't to anything special for  Python exceptions raised in the C code.

Also:
- Why do you use ctypes to call the function instead of creating an extension module?
- The way you call the function is wrong, the first argument to `PyArg_ParseTuple` should be a tuple and you're passing a string.

**Comment by ronaldoussoren** [comment]
> Thanks.
> 
> From Python docs: https://docs.python.org/3/extending/extending.html _Note The C extension interface is specific to CPython, and extension modules do not work on other Python implementations. In many cases, it is possible to avoid writing C extensions and preserve portability to other implementations. For example, if your use case is calling C library functions or system calls, you should consider using the [ctypes](https://docs.python.org/3/library/ctypes.html#module-ctypes) module or the [cffi](https://cffi.readthedocs.io/) library rather than writing custom C code. These modules let you write Python code to interface with C code and are **more portable between implementations of Python than writing and compiling a C extension module**._
> 
> Also, I trying to avoid (for now at least) modifying the path variable: _To be importable, the shared library must be available on [PYTHONPATH], and must be named after the module name, with an appropriate extension._
> 
> ctypes seemed like a low cost way to communicate with a C++ dll. It would be odd to add py_object for arguments and not support it when using it in a called function. I am using ctypes examples I have found. Perhaps they work on Linux and not Windows.

It is a low cost way to communicatie with a C/C++ shared library, but that's most commonly done for accessing an existing C API that doesn't know about the CPython API (e.g. have a C string argument instead of a PyObject that represents a Python string).

> 
> Tried your suggestion (thanks): This is from stackoverflow with your suggested change. Fails in the same way.

I haven't tried to reproduce this locally yet (on macOS), on first glance the a mismatch between the `prototype` variable and the C++ prototype of the function.

> 
> Python:
> 
> ```
> class MyClass:
>     @classmethod
>     def getAddress(cls):
>         return "Some Address"
> 
> prototype = c.PYFUNCTYPE(    
>     c.c_char_p,                
>     c.c_int, c.py_object
> )

In C++ the prototype is: ``char* (prototype)(int, PyObject*)``, the C++ code below returns a `PyObject*`.

> 
> str_path = os.path.dirname(os.path.realpath("..\\..\\string\\x64\\Debug\\string.dll"))
> handle = ctypes.PyDLL(str_path + "/string.dll")     
> 
> func = prototype(('MyFunction', handle))
> 
> q = id(MyClass)
> print(q)
> print("\n")
> func( 1, MyClass )
> ```
> 
> C++:
> 
> ```
> 	__declspec(dllexport) PyObject* MyFunction(int arg1, PyObject* obj)
> 	{
> 		PyObject* func, * res, * test;
> 
> 		//function getAddress of python object
> 		func = PyObject_GetAttrString(obj, "getAddress");
> 
> 		res = PyObject_CallFunction(func, NULL);
> 		return res;
> 	}
> ```

**Comment by gtwpro** [comment]
Thanks,

I am just taking existing examples from stackoverflow and trying to get them to work.  The address of the PyObject in c++ is the same as the object in Python.

https://stackoverflow.com/questions/77903211/call-c-function-using-python-ctypes-with-py-object-argument?noredirect=1#comment137352335_77903211

contains a picture showing the Python exception.  So, I think the issue is more fundamental than possible calling convention mismatches.  This same sort of code works fine in an extension (with self being added as a first argument).

Any piece of code simply using ctypes.py_object calling a C++ piece of code extracting a value from a PyObject * should demonstrate the issue.

**Comment by ZeroIntensity** [comment]
Yeah, as Ronald said, the problem here is that you're using `CDLL` instead of `PyDLL`. `CDLL` detaches the [thread state](https://docs.python.org/3/glossary.html#term-attached-thread-state) (formerly referred to as the GIL) around FFI calls, and it's not safe to execute most of Python's C API without an attached thread state.
